### PR TITLE
Make Pebble a child subreaper and reap zombie children

### PR DESCRIPTION
**Pull request by benhoyt**
This PR sets Pebble as a Linux "child subreaper", meaning if it's not running as PID 1, it'll get zombie (non-waited-on) children re-parented to Pebble rather than whatever *is* running as PID 1, so that we can reap them.

It also adds a "reaper" background task whose main purpose is to ensure that zombie / non-waited-on child processes are reaped (i.e., waited on and removed from the process table).

The reaper works by waiting for `SIGCHLD`, and on receipt of that signal, reaps as many waiting children as are available. For PIDs that Pebble has started directly using `os/exec`, instead of calling `cmd.Wait` we now call `reaper.WaitCommand()` -- this registers the PID with the reaper so it can notify the caller when it exited.

This is a bit klunky, but it avoids a race condition between `Wait4(-1, ...)` waiting for all PIDs, and `cmd.Wait()` waiting for a single PID -- they all go to the `Wait4(-1, ...)` now. `reaper.WaitCommand` still needs to call `cmd.Wait()` internally after the reaper has reaped the PID, so that `os/exec` cleans up goroutines and file descriptors after itself (it expects `cmd.Wait` to return a "no child processes" error).

This should fix issues like https://github.com/canonical/pebble/issues/71.

### CODE REVIEW

File: internal/overlord/servstate/handlers.go [hunk 1]
```diff
@@ -62,6 +62,9 @@ const (
 )
 
 // serviceState represents the state a service's state machine is in.
+//
+// See state-diagram.dot (and the generated state-diagram.svg image) for a
+// diagram of the states and transitions. Please try to keep these up to date!
```
Comment (benhoyt) [inline internal/overlord/servstate/handlers.go:67]: Drive-by comment addition -- should have added this in the health checks PR.

File: internal/overlord/servstate/handlers.go [hunk 2]
```diff
@@ -359,12 +363,16 @@ func (s *serviceData) startInternal() error {
...
 	}
+	logger.Debugf("Service %q started with PID %d", serviceName, s.cmd.Process.Pid)
 	s.resetTimer = time.AfterFunc(s.config.BackoffLimit.Value, func() { logError(s.backoffResetElapsed()) })
 
 	// Start a goroutine to wait for the process to finish.
 	done := make(chan struct{})
 	go func() {
 		waitErr := s.cmd.Wait()
+		if waitErr != nil {
+			logger.Debugf("Service %q Wait() error: %v", serviceName, waitErr)
```
Comment (benhoyt) [inline internal/overlord/servstate/handlers.go:374]: Add a debug log here as we didn't log the error message of this waitErr anywhere else.

File: internal/overlord/servstate/reaper.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		logger.Debugf("Reaper waiting for SIGCHLD")
+		select {
+		case <-sigChld:
+			logger.Debugf("Reaper received SIGCHLD")
+
+			// This allows ServiceManager's cmd.Wait() to pick it up before
+			// the reaper for normal service process exits, and avoid
+			// returning a "waitid: no child processes" error. A bit hacky,
+			// but I don't know another simple way to prevent this.
+			time.Sleep(25 * time.Millisecond)
```
Comment (hpidcock) [inline internal/overlord/servstate/reaper.go:57]: I think we can have a solution on this but I need to think about it more.

One possible solution is before a call to cmd.Wait, the PID is registered with the reaper, so the reaper may keep the status info if there is any. After cmd.Wait returns, it unregisters from the reaper, collecting any status it had, if the reaper has a status, use it, otherwise use the cmd.Wait's result. We'd need the reaper to have a mutex so that the register is called after cmd.Start but not before the reaper calls wait4.

**Review by hpidcock — APPROVED** [review]
LGTM but an idea around the removing of the Sleep, because it may work in your use case but not on other heavily loaded systems.

**Comment by benhoyt** [comment]
Some notes from our (Gustavo and my) discussion the other day: we definitely don't want to keep the `time.Sleep` hack, and I'm going to go back to the drawing board a bit here, finding a more general way to clean up child processes.

File: internal/overlord/servstate/handlers.go (hunk 1 above)
Comment (benhoyt) [reply internal/overlord/servstate/handlers.go:68]: Removed here and added in https://github.com/canonical/pebble/pull/105

File: internal/overlord/servstate/manager.go [hunk 4]
```diff
@@ -62,12 +66,35 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string, serv
...
+	}
+	if isSubreaper {
+		manager.stopReaper = make(chan struct{})
+		manager.reaperStopped = make(chan struct{})
+		go func() {
+			reapChildren(manager.stopReaper)
+			close(manager.reaperStopped)
+		}()
+	} else {
+		logger.Noticef("child subreaping unavailable on this platform")
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply internal/overlord/servstate/manager.go]: Done - I've made this a hard error now (it's required!) and fixed the log message style.

File: internal/overlord/servstate/manager.go [hunk 5]
```diff
@@ -62,12 +66,35 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string, serv
...
+	if err != nil {
+		return nil, fmt.Errorf("cannot set daemon as child subreaper: %w", err)
+	}
+	if isSubreaper {
+		manager.stopReaper = make(chan struct{})
+		manager.reaperStopped = make(chan struct{})
+		go func() {
+			reapChildren(manager.stopReaper)
+			close(manager.reaperStopped)
+		}()
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply internal/overlord/servstate/manager.go]: Done. It's not hugely simpler in this case, but might as well be consistent.

File: internal/overlord/servstate/reaper.go (hunk 3 above)
Comment (benhoyt) [reply internal/overlord/servstate/reaper.go]: I've implemented (something like) this now.

File: internal/overlord/servstate/reaper.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	sigChld := make(chan os.Signal, 1)
+	signal.Notify(sigChld, unix.SIGCHLD)
+	for {
+		logger.Debugf("Reaper waiting for SIGCHLD")
+		select {
+		case <-sigChld:
+			logger.Debugf("Reaper received SIGCHLD")
+
+			// This allows ServiceManager's cmd.Wait() to pick it up before
+			// the reaper for normal service process exits, and avoid
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply internal/overlord/servstate/reaper.go]: I've reworked this to avoid the non-deterministic behavior. We avoid calling `cmd.Wait` till after the reaper has reaped a registered PID, so the reaper will always win.

**Review by niemeyer — APPROVED** [review]
Nice work, Ben!  Here are some points for your consideration. As I'll be off for a few days, if you need to synchronize on these points I suggest discussing them with Harry. Once you're both happy, feel free to merge.

File: internal/reaper/reaper.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+			reapOnce()
+		case <-reaperTomb.Dying():
+			signal.Reset(unix.SIGCHLD)
+			logger.Debugf("Reaper stopped.")
+			return nil
+		}
+	}
+}
+
+// reapOnce waits for zombie child processes until there are no more.
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply internal/reaper/reaper.go]: Ah yes, thanks -- this was just a comment I hadn't updated properly.

File: internal/overlord/servstate/handlers.go [hunk 8]
```diff
@@ -345,28 +346,36 @@ func (s *serviceData) startInternal() error {
...
 	s.resetTimer = time.AfterFunc(s.config.BackoffLimit.Value, func() { logError(s.backoffResetElapsed()) })
 
 	// Start a goroutine to wait for the process to finish.
 	done := make(chan struct{})
+	cmd := s.cmd
 	go func() {
-		waitErr := s.cmd.Wait()
+		exitCode, waitErr := reaper.WaitCommand(cmd)
+		if waitErr != nil {
+			logger.Debugf("Service %q unexpected error: %v", serviceName, waitErr)
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply internal/overlord/servstate/handlers.go]: Agreed -- I updated this one to `Noticef`. Note that the log in the `else` case I've left as `Debugf` because it gets logged as Notice in almost all cases inside `exited()` (the debug log there is just to help with debugging -- showing the exit code -- for the rare branches where it doesn't).

**Comment by jnsgruk** [comment]
Nice! @marceloneppel this should land in the next Juju release I imagine (@benhoyt 2.9.29?)

**Comment by benhoyt** [comment]
Yep, that's right!

**Comment by marceloneppel** [comment]
Amazing! Thanks @benhoyt @jnsgruk and all the team involved!
